### PR TITLE
Added device setup

### PR DESCRIPTION
**Pull request by Raelr**

### CODE REVIEW

**Comment by jonjondev** [comment]
You need to tag me on the PR.

**Comment by Raelr** [comment]
@J-Mo63 since you aren't a collaborator on this repo, I thought I'd just get your review on the code itself rather than go through that entire process.

File: src/Window/Window.h [hunk 1]
```diff
@@ -1,5 +1,6 @@
 #pragma once
 
+#define GLFW_INCLUDE_VULKAN
```
Comment (jonjondev) [inline src/Window/Window.h:4]: comment this and separate from includes

File: src/Window/Window.cpp [hunk 2]
```diff
@@ -33,4 +33,9 @@ namespace SnekVk
         window = glfwCreateWindow(width, height, name, nullptr, nullptr);
```
Comment (jonjondev) [inline src/Window/Window.cpp]: This can be handled in the constructor if unused elsewhere.

File: src/Renderer/Device/VulkanDevice.h [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+namespace SnekVk {
+
+class VulkanDevice {
+	public:
+
+#ifdef DEBUG
+	const bool enableValidationLayers = true;
+#else
+    const bool enableValidationLayers = false;
+#endif
```
Comment (jonjondev) [inline src/Renderer/Device/VulkanDevice.h]:
```suggestion
#define DEBUG
	ENABLE_VALIDATION_LAYERS true
#else
        ENABLE_VALIDATION_LAYERS false
#endif

const bool enableValidationLayers = ENABLE_VALIDATION_LAYERS;
```

Consider moving macro switch out of the class with something like this - might be able to simply define with `DEBUG`.

File: src/Renderer/Device/VulkanDevice.cpp [hunk 4]
```diff
@@ -0,0 +1,362 @@
...
+namespace SnekVk {
+
+	VulkanDevice::VulkanDevice(Window &window) : window{window} 
+	{
+		CreateInstance();
+		SetupDebugMessenger();
+		CreateSurface();
+		PickPhysicalDevice();
+		CreateLogicalDevice();
+		CreateCommandPool();
```
Comment (jonjondev) [inline src/Renderer/Device/VulkanDevice.cpp:17]: very clear

File: src/Renderer/Device/VulkanDevice.cpp [hunk 5]
```diff
@@ -0,0 +1,362 @@
...
+
+	void VulkanDevice::CreateInstance() 
+	{
+		if (enableValidationLayers)
+		{
+			SNEK_ASSERT(SnekVk::CheckValidationLayerSupport(validationLayers.data(), validationLayers.size()),
+				"Validation Layers are not supported!");
+		}
+			
+		VkApplicationInfo appInfo = {};
```
Comment (jonjondev) [inline src/Renderer/Device/VulkanDevice.cpp]:
```suggestion
		VkApplicationInfo appInfo {};
```
for your consideration et. al

File: src/Renderer/Device/VulkanDevice.cpp [hunk 6]
```diff
@@ -0,0 +1,362 @@
...
+		appInfo.applicationVersion = VK_MAKE_VERSION(1, 0, 0);
+		appInfo.pEngineName = "No Engine";
+		appInfo.engineVersion = VK_MAKE_VERSION(1, 0, 0);
+		appInfo.apiVersion = VK_API_VERSION_1_0;
+
+		VkInstanceCreateInfo createInfo = {};
+		createInfo.sType = VK_STRUCTURE_TYPE_INSTANCE_CREATE_INFO;
+		createInfo.pApplicationInfo = &appInfo;
+
+		auto extensions = SnekVk::GetRequiredExtensions(enableValidationLayers);
```
Comment (jonjondev) [inline src/Renderer/Device/VulkanDevice.cpp]: Throw this in a namespace for clarity

File: src/Renderer/Device/VulkanDevice.cpp [hunk 7]
```diff
@@ -0,0 +1,362 @@
...
+		for (uint32_t i = 0; i < memProperties.memoryTypeCount; i++) 
+		{
+			if ((typeFilter & (1 << i)) &&
+				(memProperties.memoryTypes[i].propertyFlags & properties) == properties) 
+			{
+			return i;
+			}
+		}
+
+		throw std::runtime_error("failed to find suitable memory type!");
```
Comment (jonjondev) [inline src/Renderer/Device/VulkanDevice.cpp:219]: Assert?

File: src/Renderer/Renderer.h [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+typedef int32_t i32;
+typedef size_t size;
+
+#if (defined(_WIN32) || defined(_WIN64)) 
+    #define EXIT_APP \
+        system("pause"); \
+        abort();
+#else
+    #define EXIT_APP abort();
+#endif
```
Comment (jonjondev) [inline src/Renderer/Renderer.h:18]: MULTILEVEL_MACRO_MADNESS ANOTHER \
     MACRO \
     INSIDE!

File: src/Renderer/Renderer.h [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+    } \
+
+namespace SnekVk 
+{
+    enum SnekState 
+    {
+        Success = 0,
+        Failure = 1
+    };
+}
```
Comment (jonjondev) [inline src/Renderer/Renderer.h]: iu

**Review by jonjondev — APPROVED** [review]
Awful. Just awful. Let's merge.

File: src/Renderer/Renderer.h (hunk 8 above)
Comment (Raelr) [reply src/Renderer/Renderer.h:18]: Fight me.

File: src/Renderer/Device/VulkanDevice.cpp (hunk 6 above)
Comment (Raelr) [reply src/Renderer/Device/VulkanDevice.cpp]: Done.

File: src/Renderer/Device/VulkanDevice.h (hunk 3 above)
Comment (Raelr) [reply src/Renderer/Device/VulkanDevice.h]: Changed.

**Pull request merged**